### PR TITLE
doc: Documents server only Auth method updateUserById()

### LINKED ISSUE #5115: Docs: Supabase Auth (Server Only) Missing updateUserById reference docs
# Improve documentation

The Auth (Server Only) docs at https://supabase.com/docs/reference/javascript/auth-api-createuser is missing an entry for:

```ts
  /**
   * Updates the user data.
   *
   * @param attributes The data you want to update.
   *
   * This function should only be called on a server. Never expose your `service_role` key in the browser.
   */
  async updateUserById(
    uid: string,
    attributes: AdminUserAttributes
  ): Promise<{ user: User | null; data: User | null; error: ApiError | null }> {
```

See: https://github.com/supabase/gotrue-js/blob/edcd2ffde6dfe81d02700a52a79e635a5ab8839d/src/GoTrueApi.ts#L561

## Link 

Add a link to the page which needs improvement (if relevant): 

* https://supabase.com/docs/reference/javascript/auth-api-createuser

* Method is: https://github.com/supabase/gotrue-js/blob/edcd2ffde6dfe81d02700a52a79e635a5ab8839d/src/GoTrueApi.ts#L561

## Describe the problem

This method can be used to update (server-side) a user's app_metadata, using a service role key, which was not possible prior to gotrue-js 1.22.

### PR DESCRIPTION
**Pull request by dthyresson**
## What kind of change does this PR introduce?

Fixes https://github.com/supabase/supabase/issues/5115

Documents server only Auth method `updateUserById()` which was introduced in https://github.com/supabase/gotrue-js/pull/195

Note that GotTrue v1.22 has this method, but is not yet released as part of the Supabase client.

## Additional context

Also:

* Missing properties like email, user_metadata for `AdminUserAttributes`. Is TS up to date? See: https://supabase.github.io/gotrue-js/spec.json
* Note: I noticed @kiwicopple and @kangmingtay that `UserAttributes` allows a user to change their own `email_change_token`. Is this the case? Or should it be moved into `AdminUserAttributes`?

### CODE REVIEW

File: web/sidebar_spec_supabase.js [hunk 1]
```diff
@@ -9,13 +9,13 @@ module.exports = {
     {
       type: 'category',
       label: 'Auth',
-      items: ['reference/javascript/auth-signup', 'reference/javascript/auth-signin', 'reference/javascript/auth-signout', 'reference/javascript/auth-session', 'reference/javascript/auth-user', 'reference/javascript/auth-update', 'reference/javascript/auth-setauth', 'reference/javascript/auth-onauthstatechange', 'reference/javascript/auth-api-getuser', 'reference/javascript/reset-password-email'],
+      items: ['reference/javascript/auth-signup', 'reference/javascript/auth-signin', 'reference/javascript/auth-signout', 'reference/javascript/auth-session', 'reference/javascript/auth-user', 'reference/javascript/auth-update', 'reference/javascript/auth-setauth', 'reference/javascript/auth-onauthstatechange', 'reference/javascript/auth-api-getuser'],
```
Comment (kiwicopple) [inline web/sidebar_spec_supabase.js:12]: @dthyresson - any reason why this one was removed? `reference/javascript/reset-password-email'],`

I believe this is required for "Reset password" functionality?

Comment (dthyresson) [reply web/sidebar_spec_supabase.js:12]: That was an accident as I thought it's details were duplicated in the yaml (there were multiple heading info sections for that method, I think). Thanks for catching. I'll restore it.

Comment (dthyresson) [reply web/sidebar_spec_supabase.js:12]: Oh I see my error - I mistook it for a duplicate of the Auth Server only entry.

**Comment by dthyresson** [comment]
I think removing the entry was ok to do, but we can check on the Vercel preview deploy if approved.

I also added an alias for starting the server since after a build:


```term
Component TabItem was not imported, exported, or provided by MDXProvider as global scope
Success! Generated static files in build.

Use `npm run serve` to test your build locally.
```

`npm run serve`

is suggested, but that command doesn't exist -- and now is an alias to start Docusaurus.

File: web/sidebar_spec_supabase.js (hunk 1 above)
Comment (awalias) [reply web/sidebar_spec_supabase.js:12]: @dthyresson does this one still need resolving?

**Comment by dthyresson** [comment]
I think we’re ok on this now, though is it possible to preview the deployed site to review the docs pages? When I look there are no styles and there is a warning.

**Comment by kiwicopple** [comment]
> I think we’re ok on this now, though is it possible to preview the deployed site to review the docs pages? When I look there are no styles and there is a warning.

I see you resolved [this](https://github.com/supabase/supabase/pull/5140/files#r791668756) one, but I'm pretty certain we still want that function in the docs right? Any reason why we would remove it from the clientside functions?

**Comment by dthyresson** [comment]
> I'm pretty certain we still want that function in the docs right? Any reason why we would remove it from the clientside functions?

I'm sorry, I must not be understanding something, but, there are the documented functions:

image


image


This method is admin only and needs a service role key -- so, it is not avialoblae on the client side.

**Comment by kiwicopple** [comment]
> This method is admin only and needs a service role key -- so, it is not available on the client side.

is this 100% confirmed? As far as I can remember it works client side, and doesn't need a service key. And it shouldn't either right? Otherwise a developer _needs_ to implement server functions for a "Forgot password" button.

I think I had some old tests on this one in gotrue-js. Perhaps the confusion is because the function is in the `GoTrueApi.js` file? (which has both server and client routes)

**Comment by dthyresson** [comment]
> > This method is admin only and needs a service role key -- so, it is not available on the client side.
> 
> is this 100% confirmed? As far as I can remember it works client side, and doesn't need a service key. 

Ah, now I understand finally. Thanks!

The GoTrue api endpoint:

https://github.com/supabase/gotrue/blob/e67a10c12e8aac3d9a56ae6aff167ac0e5431670/api/api.go#L120

is not and Admin-only endpoint, despite the docs.

And the `/recover` endpoint takes an `email`:

https://github.com/supabase/gotrue/blob/e67a10c12e8aac3d9a56ae6aff167ac0e5431670/api/recover.go#L18

So, I do need to restore:

```
  Reset Password (Email):
    $ref: '@supabase/gotrue-js.GoTrueApi.resetPasswordForEmail'
    notes: |
      Sends a reset request to an email address.
      When the user clicks the reset link in the email they will be forwarded to:
      `<SITE_URL>#access_token=x&refresh_token=y&expires_in=z&token_type=bearer&type=recovery`
      Your app must detect `type=recovery` in the fragment and display a password reset form to the user.
      You should then use the access_token in the url and new password to update the user as follows:
      ```js
      const { error, data } = await supabase.auth.api
            .updateUser(access_token, { password : new_password })
      ```
    examples:
      - name: Reset password
        isSpotlight: true
        js: |
          ```js
          const { data, error } = supabase.auth.api
            .resetPasswordForEmail('user@email.com')
```

as part of the "client" docs even though it uses `GoTrueApi` vs `GoTrueClient`.

And update the Server only docs as well.

**Comment by dthyresson** [comment]
Ok, somehow the `make` and `run npm build` scripts force this structure as it generates the sidebar:

```
    {
      type: 'category',
      label: 'Auth',
      items: ['reference/javascript/auth-signup', 'reference/javascript/auth-signin', 'reference/javascript/auth-signout', 'reference/javascript/auth-session', 'reference/javascript/auth-user', 'reference/javascript/auth-update', 'reference/javascript/auth-setauth', 'reference/javascript/auth-onauthstatechange', 'reference/javascript/auth-api-getuser'],
      collapsed: true,
    },
    {
      type: 'category',
      label: 'Auth (Server Only)',
      items: ['reference/javascript/auth-api-createuser', 'reference/javascript/auth-api-deleteuser', 'reference/javascript/auth-api-generatelink', 'reference/javascript/auth-api-inviteuserbyemail', 'reference/javascript/auth-api-sendmobileotp', 'reference/javascript/auth-api-resetpasswordforemail', 'reference/javascript/auth-api-updateuserbyid'],
      collapsed: true,
    },
```

Where `'reference/javascript/auth-api-resetpasswordforemail'` is always assigned to:

```
      type: 'category',
      label: 'Auth (Server Only)',
```

If I manually update the `sidebar_spec_supabase` content is overwritten to above.

Will look into the .json specs that perhaps it pulls in to generate.

**Comment by kiwicopple** [comment]
OK nearly there :)

I see some of the routes have changed now, so we just need to add redirects into this file:
https://github.com/supabase/supabase/blob/master/www/next.config.js

**Comment by dthyresson** [comment]
> OK nearly there :)
> 
> I see some of the routes have changed now, so we just need to add redirects into this file: https://github.com/supabase/supabase/blob/master/www/next.config.js

Ok will do. Thanks for patience as I got up to speed on the doc generators and processes.

**Comment by dthyresson** [comment]
Looks like this redirect needs be added from old:

image


`docs/reference/javascript/reset-password-email` -> `docs/reference/javascript/auth-api-resetpasswordforemail`

**Review by kiwicopple — APPROVED** [review]

**Comment by kiwicopple** [comment]
thanks DT!

**Pull request merged**